### PR TITLE
feat: add example apps

### PR DESCRIPTION
**Pull request by jsnwesson**
[APER-3042](https://2u-internal.atlassian.net/browse/APER-3042)

### This PR includes: 
- example apps that display the use of direct and iFrame plugins
- examples of using all four Plugin Operations (Insert, Modify, Wrap, Hide)
- a bug fix when using `getConfig` in usePluginSlot
- update how Plugin and PluginSlot retrieve `intl` using `useIntl` function from `frontend platform`

### Steps to run example apps
Run `npm install` in the root directory, then

Using two terminals:
1. Run `npm install` inside both example app directories. 
2. Run `npm run start` in both directories.

Alternatively, once the packages are installed in both apps, you can run the apps from the root directory. 
1.`npm run start` runs the host MFE (example) 
2. `npm run start:plugins` runs the child MFE (example-plugin-app)

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export {
+  Plugin,
+  PluginSlot,
+  PLUGIN_OPERATIONS,
+  IFRAME_PLUGIN,
+  DIRECT_PLUGIN,
+} from './plugins';
```
Comment (justinhynes) [inline src/index.js:7]: Oh boy, here is a dumb question. 

What's the difference between a `index.jsx` and `index.js`? Why not just update the existing `index.jsx`?v 😬

File: README.rst [hunk 2]
```diff
@@ -26,11 +26,23 @@ for a component that lives in another MFE (the 'Child MFE') to be plugged into a
...
+Add Library Dependency
 -------------------------
 
 Add ``@edx/frontend-plugin-framework`` to the ``package.json`` of both Host and Child MFEs.
 
+Using the Example Apps
+----------------------
+
+To use the example apps, you will need to open two terminals.
+1. Run ``npm install`` inside both example app directories.
```
Comment (justinhynes) [inline README.rst:38]: Can we add some more detail here? Maybe the path to the explicit directories? "inside both example app directories" may make more sense to the folks who have been doing the work... but I'm not really sure what this means?

File: example/env.config.jsx [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    <p>With this wrapper, you can add anything before or after the widget.</p>
+  </div>
+);
+
+// Note that in an actual application this file would be added to .gitignore.
+const config = {
+  JS_FILE_VAR: 'JS_FILE_VAR_VALUE_FOR_EXAMPLE_APP',
+  EXAMPLE_VAR: 'Example Value',
+  ACCESS_TOKEN_COOKIE_NAME: 'edx-jwt-cookie-header-payload',
+  ACCOUNT_PROFILE_URL: 'http://localhost:1995',
```
Comment (justinhynes) [inline example/env.config.jsx:34]: It looks like the default config has entries not found in `example-plugin-app/env.config.js`. Should the default configuration match between these two files (e.g. `ACCOUNT_PROFILE_URL` appears here, but not in the other example/default file). 

Does it matter that the ports defined in the `BASE_URL` between the two files are different?

File: example/package.json [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  "description": "",
+  "main": "index.jsx",
+  "scripts": {
+    "lint": "fedx-scripts eslint --ext .js --ext .jsx . --resolve-plugins-relative-to .",
+    "lint:fix": "fedx-scripts eslint --fix --ext .js --ext .jsx . --resolve-plugins-relative-to .",
+    "start": "fedx-scripts webpack-dev-server --progress"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
```
Comment (justinhynes) [inline example/package.json:13]: Not sure if this matters at all. I noticed that most other licenses in other frontend/MFE repos is set to `AAGPL-3.0`. Should this match? Not sure what the consequences are, if any, for a mismatch.

File: example-plugin-app/package.json [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    "start": "fedx-scripts webpack-dev-server --progress"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "@openedx/frontend-build": "^13.0.28",
+    "core-js": "^3.29.1",
+    "react": "^16.14.0",
+    "react-dom": "^16.14.0",
```
Comment (MaxFrank13) [inline example-plugin-app/package.json:18]: Do we want to keep these React versions up to date with the MFEs?

File: example/env.config.jsx [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            RenderWidget: ModularDirectPlugin,
+            content: {
+              title: 'Inserted Direct Plugin',
+              uniqueText: 'This is a direct plugin with priority of 1, which is why it appears first in this slot.',
+            },
+          },
+        },
+      ],
+    },
+  },
```
Comment (MaxFrank13) [inline example/env.config.jsx:168]: How do you feel about putting each of the configurations for the `pluginSlots` object in their own file? Something like `pluginSlots.config.js(x)` that exports an object for each slot. Is there a downside to that? I could see it making things clearer for devs as well as establishing a cleaner pattern for them to follow in their own implementations

File: example/package.json [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  "dependencies": {
+    "core-js": "^3.29.1",
+    "prop-types": "^15.8.1",
+    "react": "^16.14.0",
+    "react-dom": "^16.14.0",
```
Comment (MaxFrank13) [inline example/package.json:18]: Same question as the other package.json. Should these be React 17 at least to conform with the MFEs?

File: example/src/directPlugins/DefaultDirectWidget.jsx [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+
+export default function DefaultDirectWidget() {
+  return (
+    <section className="bg-success p-3 text-light">
+      <h3>Default Direct Widget</h3>
+      <p>
+        {`This widget is a default component that lives in the example app and is directly inserted via JS configuration.
+        Note that this default widget appears after the Inserted Direct Plugin. This is because this component's
+        "priority" is set to 20 in the JS config.`}
```
Comment (MaxFrank13) [inline example/src/directPlugins/DefaultDirectWidget.jsx:10]: Nit: does this need to be a template string or can it just be written in the markup?

File: example/src/directPlugins/ModularDirectPlugin.jsx [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+import PropTypes from 'prop-types';
+
+export default function ModularDirectPlugin({ content }) {
+  return (
+    <section className="bg-light p-3">
+      <h3>{ content.title }</h3>
+      <p>
+        {`This plugin is a component that lives in the example app and is directly inserted via JS configuration.
+        What makes this plugin unique is that it takes in some "content" from the JS configuration and is able
+        to apply that content accordingly (as shown below).`}
```
Comment (MaxFrank13) [inline example/src/directPlugins/ModularDirectPlugin.jsx:11]: Nit: same as previous, can this just be written normally?

File: example/src/directPlugins/PluginDirect.jsx [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+
+export default function PluginDirect() {
+  return (
+    <section className="bg-warning p-3">
+      <h3>Inserted Direct Plugin</h3>
+      <p>
+        {`This plugin is a component that lives in the example app and is directly inserted via JS configuration.
+        What makes this unique is that it isn't part of the default content defined for this slot, but is instead
+        inserted as a plugin slot change "PLUGIN_OPERATIONS.Insert".`}
```
Comment (MaxFrank13) [inline example/src/directPlugins/PluginDirect.jsx:10]: Nit: template string vs regular string

File: src/index.js (hunk 1 above)
Comment (MaxFrank13) [reply src/index.js:7]: The `.jsx` will tell the compiler and linter what to expect in the file (in this case, React code). It looks like Jason has removed the `index.jsx` that used to live in `src` (it had old example code that came with the template). 

Now, the `index` in src doesn't need to be `jsx` because it's just exporting all of the components from `frontend-plugin-framework` — prior to this PR, that was happening in `src/plugins/index.js`. 

[extra inform] In MFEs, the index in `src` _does_ need to be `jsx` because that's where the app in initialized, the router is setup, and other React code is fired.

File: src/plugins/PluginContainerIframe.jsx [hunk 11]
```diff
@@ -31,9 +31,9 @@ export const IFRAME_FEATURE_POLICY = (
   'fullscreen; microphone *; camera *; midi *; geolocation *; encrypted-media *'
 );
 
-const PluginContainerIframe = ({
+function PluginContainerIframe({
```
Comment (MaxFrank13) [inline src/plugins/PluginContainerIframe.jsx:34]: Was the linter complaining about this and that's why you made the change or was there some other reason? I don't have a request to change it back, just curious why this change was made

File: src/plugins/PluginSlot.jsx [hunk 12]
```diff
@@ -2,26 +2,23 @@ import React, { forwardRef } from 'react';
...
+import messages from './Plugin.messages';
 import { usePluginSlot } from './data/hooks';
 import PluginContainer from './PluginContainer';
 import { organizePlugins, wrapComponent } from './data/utils';
 
 const PluginSlot = forwardRef(({
-  as, id, intl, pluginProps, ...props
+  as, id, pluginProps, ...props
 }, ref) => {
-  /** TODO: Examples still need to be set up as part of APER-3042 https://2u-internal.atlassian.net/browse/APER-3042 */
```
Comment (MaxFrank13) [inline src/plugins/PluginSlot.jsx:20]: 🎉

File: src/plugins/PluginContainerIframe.jsx (hunk 11 above)
Comment (jsnwesson) [reply src/plugins/PluginContainerIframe.jsx:34]: ah yeah, I rant into these lint problems once I got the `.eslintrc` config set up properly, so instead of ignoring them I decided it'd be better to make everything uniform going forward.

File: example/src/directPlugins/ModularDirectPlugin.jsx (hunk 9 above)
Comment (jsnwesson) [reply example/src/directPlugins/ModularDirectPlugin.jsx:11]: it was a lint thing with using quotations without wrapping them with `{` `}`. I can look into writing it out so it doesn't need template strings, I admit it's a bit odd to have when it isn't being used for string variables

File: example/env.config.jsx (hunk 6 above)
Comment (jsnwesson) [reply example/env.config.jsx:168]: I thought about how we could make this cleaner, too, but I think that until we make a decision on how to make that look realistic to what is needed in Production, this could serve as a first pass.

To just get my thoughts out:
For one, I know that we'll be aiming to remove the `defaultContents` logic so it shouldn't live in a JS config, so that removes half the bloat. 
I don't think there's an immediate downside to your suggestion, but we'd need to make sure that there's an `env.config.js` that knows to import those slots; in production, that would mean doing something like having two JS config files that get copied into a repo, and that can start to look weird.

File: example/package.json (hunk 4 above)
Comment (jsnwesson) [reply example/package.json:13]: I'm not sure it matters since they're example apps within a repo that already has a license, BUT I can change them to AAGPL-3.0 for conformity sake, thanks for the catch!

File: example/env.config.jsx (hunk 3 above)
Comment (jsnwesson) [reply example/env.config.jsx:34]: Huh! Not sure how they ended up becoming different, but I'll fix that. For the `BASE_URL`, since the two apps are meant to act as different MFEs running on different URL/ports, they should be different.

File: example/src/directPlugins/ModularDirectPlugin.jsx (hunk 9 above)
Comment (jsnwesson) [reply example/src/directPlugins/ModularDirectPlugin.jsx]: Pushed a change to remove this format and opted for using `&quot;` or `&apos;` when needed

File: example/package.json (hunk 7 above)
Comment (jsnwesson) [reply example/package.json:18]: That's a fair point! Bumping these versions to stay with the root's package.json version of React

File: example/package.json (hunk 4 above)
Comment (jsnwesson) [reply example/package.json]: Done!

File: README.rst (hunk 2 above)
Comment (jsnwesson) [reply README.rst]: More details added!

**Review by MaxFrank13 — APPROVED** [review]

**Review by justinhynes — APPROVED** [review]
